File: src/placeholders.ts
// Placeholders: {{name}} / {{name=default}} (spec §2). Anything else stays
// literal; substitution is verbatim — template authors do their own quoting.

export interface Placeholder {
  name: string;
  default?: string;
}

const PLACEHOLDER_RE = /\{\{([A-Za-z0-9_-]+)(?:=([^}]*))?\}\}/g;

// Unique by name (repeats prompt once); first default wins.
export function parsePlaceholders(template: string): Placeholder[] {
  const seen = new Map<string, Placeholder>();
  for (const m of template.matchAll(PLACEHOLDER_RE)) {
    const name = m[1]!;
    if (!seen.has(name)) seen.set(name, { name, default: m[2] });
  }
  return [...seen.values()];
}

// Value for a name: supplied value > first-wins default > empty — every
// occurrence of a repeated name renders identically.
function resolveValues(template: string, values: Record<string, string>): Record<string, string> {
  const resolved: Record<string, string> = {};
  for (const p of parsePlaceholders(template)) resolved[p.name] = values[p.name] ?? p.default ?? '';
  return resolved;
}

export function renderCommand(template: string, values: Record<string, string>): string {
  const resolved = resolveValues(template, values);
  return template.replace(PLACEHOLDER_RE, (_, name: string) => resolved[name]!);
}

// Split into literal/placeholder segments with the raw {{...}} tokens kept,
// so the edit screen can highlight the Placeholder slots themselves.
export function templateSegments(template: string): Array<{ text: string; placeholder: boolean }> {
  const out: Array<{ text: string; placeholder: boolean }> = [];
  let last = 0;
  for (const m of template.matchAll(PLACEHOLDER_RE)) {
    if (m.index > last) out.push({ text: template.slice(last, m.index), placeholder: false });
    out.push({ text: m[0], placeholder: true });
    last = m.index + m[0].length;
  }
  if (last < template.length) out.push({ text: template.slice(last), placeholder: false });
  return out;
}

// Split into literal/substituted segments so the live preview can highlight
// the substituted values (spec §3.2).
export function renderSegments(
  template: string,
  values: Record<string, string>,
): Array<{ text: string; substituted: boolean }> {
  const resolved = resolveValues(template, values);
  const out: Array<{ text: string; substituted: boolean }> = [];
  let last = 0;
  for (const m of template.matchAll(PLACEHOLDER_RE)) {
    if (m.index > last) out.push({ text: template.slice(last, m.index), substituted: false });
    out.push({ text: resolved[m[1]!]!, substituted: true });
    last = m.index + m[0].length;
  }
  if (last < template.length) out.push({ text: template.slice(last), substituted: false });
  return out;
}

File: internal/placeholders/placeholders.go
// Package placeholders handles {{name}} / {{name=default}} (spec §2).
// Anything else stays literal; substitution is verbatim — template authors do
// their own quoting.
package placeholders

import "regexp"

type Placeholder struct {
	Name       string
	Default    string
	HasDefault bool
}

var re = regexp.MustCompile(`\{\{([A-Za-z0-9_-]+)(?:=([^}]*))?\}\}`)

// Parse returns the Placeholders unique by name (a repeat prompts once);
// the first default wins.
func Parse(template string) []Placeholder {
	seen := map[string]bool{}
	out := []Placeholder{}
	for _, m := range re.FindAllStringSubmatchIndex(template, -1) {
		name := template[m[2]:m[3]]
		if seen[name] {
			continue
		}
		seen[name] = true
		p := Placeholder{Name: name}
		if m[4] >= 0 {
			p.Default = template[m[4]:m[5]]
			p.HasDefault = true
		}
		out = append(out, p)
	}
	return out
}

// resolve gives every name its value: supplied > first-wins default > empty,
// so every occurrence of a repeated name renders identically.
func resolve(template string, values map[string]string) map[string]string {
	resolved := map[string]string{}
	for _, p := range Parse(template) {
		if value, ok := values[p.Name]; ok {
			resolved[p.Name] = value
		} else {
			resolved[p.Name] = p.Default
		}
	}
	return resolved
}

func Render(template string, values map[string]string) string {
	resolved := resolve(template, values)
	out := ""
	last := 0
	for _, m := range re.FindAllStringSubmatchIndex(template, -1) {
		out += template[last:m[0]] + resolved[template[m[2]:m[3]]]
		last = m[1]
	}
	return out + template[last:]
}

// Segment is a run of the template, flagged as either literal or the thing
// the caller wants highlighted.
type Segment struct {
	Text string
	Flag bool
}

// TemplateSegments splits into literal/placeholder runs with the raw {{...}}
// tokens kept, so the edit screen can highlight the Placeholder slots.
func TemplateSegments(template string) []Segment {
	out := []Segment{}
	last := 0
	for _, m := range re.FindAllStringSubmatchIndex(template, -1) {
		if m[0] > last {
			out = append(out, Segment{Text: template[last:m[0]]})
		}
		out = append(out, Segment{Text: template[m[0]:m[1]], Flag: true})
		last = m[1]
	}
	if last < len(template) {
		out = append(out, Segment{Text: template[last:]})
	}
	return out
}

// RenderSegments splits into literal/substituted runs so the live preview can
// highlight the substituted values (spec §3.2).
func RenderSegments(template string, values map[string]string) []Segment {
	resolved := resolve(template, values)
	out := []Segment{}
	last := 0
	for _, m := range re.FindAllStringSubmatchIndex(template, -1) {
		if m[0] > last {
			out = append(out, Segment{Text: template[last:m[0]]})
		}
		out = append(out, Segment{Text: resolved[template[m[2]:m[3]]], Flag: true})
		last = m[1]
	}
	if last < len(template) {
		out = append(out, Segment{Text: template[last:]})
	}
	return out
}
